pkg/operation/botanist/constraints_check.go

```diff
@@ -0,0 +1,146 @@
+	}
+
+	for _, webhookConfig := range validatingWebhookConfigs.Items {
+		if exitConstraint := checker.CheckWebhooks(constraint, "ValidatingWebhookConfiguration", webhookConfig.Name, webhookConfig.Webhooks); exitConstraint != nil {
+			return exitConstraint, nil
+		}
+	}
+
+	mutatingWebhookConfigs := &admissionregistrationv1beta1.MutatingWebhookConfigurationList{}
+	if err := b.K8sShootClient.Client().List(ctx, mutatingWebhookConfigs); err != nil {
+		return nil, fmt.Errorf("could not get MutatingWebhookConfigurations of Shoot cluster to check if Shoot can be hibernated")
+	}
+
+	for _, webhookConfig := range mutatingWebhookConfigs.Items {
+		if exitConstraint := checker.checkMutatingWebhookConfig(constraint, webhookConfig); exitConstraint != nil {
+			return exitConstraint, nil
+		}
+	}
+
+	c := gardencorev1alpha1helper.UpdatedCondition(constraint, gardencorev1alpha1.ConditionTrue, "NoProblematicWebhooks", "Shoot can be hibernated.")
+	return &c, nil
+}
+
+func (c *ConstraintsChecker) checkValidatingWebhookConfig(constraint gardencorev1alpha1.Condition, webhookConfig admissionregistrationv1beta1.ValidatingWebhookConfiguration) *gardencorev1alpha1.Condition {
+	return c.CheckWebhooks(constraint, "ValidatingWebhookConfiguration", webhookConfig.Name, webhookConfig.Webhooks)
+}
+
+func (c *ConstraintsChecker) checkMutatingWebhookConfig(constraint gardencorev1alpha1.Condition, webhookConfig admissionregistrationv1beta1.MutatingWebhookConfiguration) *gardencorev1alpha1.Condition {
+	return c.CheckWebhooks(constraint, "MutatingWebhookConfiguration", webhookConfig.Name, webhookConfig.Webhooks)
+}
+
+// CheckWebhooks checks all webhooks in the Shoot to determine if hibernation is not possible due to problematic Webhooks.
+// Problematic webhooks are: ValidationWebhookConfigurations and MutatingWebhookConfigurations with rules
+// for CREATE pods or nodes and failurePolicy=Fail and webhooks[].clientConfig.url == nil.
+// Otherwise, we can never wake up this shoot cluster again as new nodes cannot get created,
+// or our system component pods cannot get created (because the validating webhook pod is not yet running).
+func (c *ConstraintsChecker) CheckWebhooks(constraint gardencorev1alpha1.Condition, configKind string, configName string, webhooks []admissionregistrationv1beta1.Webhook) *gardencorev1alpha1.Condition {
+	for _, webhook := range webhooks {
+		if webhook.FailurePolicy != nil && *webhook.FailurePolicy != admissionregistrationv1beta1.Fail {
+			// in admissionregistration.k8s.io/v1 FailurePolicy is defaulted to `Fail`
+			// see https://github.com/kubernetes/api/blob/release-1.16/admissionregistration/v1/types.go#L195
+			// and https://github.com/kubernetes/api/blob/release-1.16/admissionregistration/v1/types.go#L324
+			// therefore, don't ignore FailurePolicy==nil
+			continue
+		}
+
+		if webhook.ClientConfig.URL != nil && len(*webhook.ClientConfig.URL) > 0 {
+			// Webhooks with external URL (out-of-cluster) are not problematic
```

Comment:
I don't know, this check is too inaccurate. I can have webhook with a url pointing to `https://my-svc.my-ns` which is still running in my cluster. Can we either improve it or get rid of it?

Reply:
Yes, this is a very rudimentary check (just as discussed in #1575) and relies on the user acting according to the [Kubernetes API docs](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#webhookclientconfig-v1-admissionregistration-k8s-io):
> url: [...] The `host` should not refer to a service running in the cluster; use the `service` field instead.

One way to do a more accurate check would probably be to explicitly test if the host of the url can be resolved externally (like `nslookup host 8.8.8.8`). 
Though, then the question would be which DNS server to test against. So this would come with some effort.

However, external webhooks with `failurePolicy=Fail` can still cause the wakeup to fail. So I would suggest to drop this check and consider external Webhooks with `failurePolicy=Fail` to also be problematic.

WDYT?

cc @vpnachev

Reply:
+1 for dropping, makes sense.